src/metabase/sync/analyze/classifiers/text_fingerprint.clj

```diff
@@ -8,45 +8,58 @@
             [metabase.util.schema :as su]
             [schema.core :as s]))
 
-(def ^:private ^:const ^Float percent-valid-threshold
+(def ^:private ^:const ^Double percent-valid-threshold
   "Fields that have at least this percent of values that are satisfy some predicate (such as `u/email?`)
    should be given the corresponding special type (such as `:type/Email`)."
   0.95)
 
-(s/defn ^:private percent-key-below-threshold? :- s/Bool
+(def ^:private ^Double lower-percent-valid-threshold
+  "Fields that have at least this lower percent of values that satisfy some predicate (such as `u/state?`) should be
+  given the corresponding special type (such as `:type/State`)"
+  0.7)
+
+(s/defn ^:private percent-key-above-threshold? :- s/Bool
   "Is the value of PERCENT-KEY inside TEXT-FINGERPRINT above the `percent-valid-threshold`?"
-  [text-fingerprint :- i/TextFingerprint, percent-key :- s/Keyword]
+  [^Double threshold, text-fingerprint :- i/TextFingerprint, percent-key :- s/Keyword]
   (boolean
    (when-let [percent (get text-fingerprint percent-key)]
-     (>= percent percent-valid-threshold))))
-
+     (>= percent threshold))))
 
 (def ^:private percent-key->special-type
   "Map of keys inside the `TextFingerprint` to the corresponding special types we should mark a Field as if the value of
   the key is over `percent-valid-thresold`."
-  {:percent-json  :type/SerializedJSON
-   :percent-url   :type/URL
-   :percent-email :type/Email})
+  {:percent-json  [:type/SerializedJSON percent-valid-threshold]
+   :percent-url   [:type/URL            percent-valid-threshold]
+   :percent-email [:type/Email          percent-valid-threshold]
+   :percent-state [:type/State          lower-percent-valid-threshold]})
 
 (s/defn ^:private infer-special-type-for-text-fingerprint :- (s/maybe su/FieldType)
   "Check various percentages inside the TEXT-FINGERPRINT and return the corresponding special type to mark the Field
   as if the percent passes the threshold."
   [text-fingerprint :- i/TextFingerprint]
-  (some (fn [[percent-key special-type]]
-          (when (percent-key-below-threshold? text-fingerprint percent-key)
+  (some (fn [[percent-key [special-type threshold]]]
+          (when (percent-key-above-threshold? threshold text-fingerprint percent-key)
             special-type))
-        (seq percent-key->special-type)))
+        percent-key->special-type))
+
+(defn- can-edit-special-type?
```

Comment:
Can you document/explain the algorithm (tagging via metadata) here

Reply:
good catch! will do